nixos/modules/services/networking/zeronsd.nix

```diff
@@ -0,0 +1,119 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.zeronsd;
+  settingsFormat = pkgs.formats.json {};
+  inherit (lib) mkOption;
+in {
+  options.services.zeronsd.enable = lib.mkEnableOption (lib.mdDoc "ZeroNSD");
+
+  options.services.zeronsd.servedNetworks = mkOption {
+    default = [];
+    example = [ "a8a2c3c10c1a68de" ];
+    type = lib.types.listOf lib.types.str;
+    description = lib.mdDoc "List of ZeroTier Network IDs to start zeronsd instance for.";
+  };
+
+  options.services.zeronsd.settings = lib.mkOption {
+    description = ''
+      Settings for zeronsd
+    '';
+    type = lib.types.submodule {
+      freeformType = settingsFormat.type;
+
+      options.domain = lib.mkOption {
+        default = "home.arpa";
+        type = lib.types.str;
+        description = lib.mdDoc ''
+          Domain under which ZeroTier records will be available
+        '';
+      };
+
+      options.token = lib.mkOption {
+        default = null;
+        type = lib.types.path;
+        description = lib.mdDoc ''
+          Path to a file containing the API Token for ZeroTier Central
+        '';
+      };
+
+      options.log-level = lib.mkOption {
+        default = "info";
+        type = lib.types.str;
+        description = lib.mdDoc ''
+          Log Level (one of off, error, warn, info, debug, trace)
+        '';
+      };
+
+      options.wildcard = lib.mkOption {
+        default = false;
+        type = lib.types.bool;
+        description = lib.mdDoc ''
+          Whether to server a wildcard record for ZeroTier Nodes
+        '';
+      };
+    };
+  };
+
+  options.services.zeronsd.package = lib.mkOption {
+    default = pkgs.zeronsd;
+    defaultText = lib.literalExpression "pkgs.zeronsd";
+    type = lib.types.package;
+    description = lib.mdDoc ''
+      Zeronsd package to use.
+    '';
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = config.services.zerotierone.enable;
+        message = "zeronsd needs a configured zerotier-one";
+      }
+      {
+        assertion = cfg.servedNetworks != [];
+        message = "At least one ZeroTier network needs to be configured in services.zeronsd.servedNetworks";
+      }
+    ];
+    systemd.services = builtins.listToAttrs (builtins.map (net: {
+      name = "zeronsd-${net}";
+      value = {
+        description = "ZeroTier DNS server for Network ${net}";
+
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" "zerotierone.service" ];
+        requires = [ "network-online.target" ];
```

Comment:
Not sure if this is a good idea as that would stop the service if the network is (partially) not online